### PR TITLE
Diag

### PR DESCRIPTION
**Pull request by alewis**
Replaces diag with diagflat and diagonal. Modifies trace to use the same call signature as numpy, and therefore to support batching.

### CODE REVIEW

File: tensornetwork/backends/jax/jax_backend_test.py [hunk 1]
```diff
@@ -787,3 +769,49 @@ def A_mv(x):
   rtol = tol*jax.numpy.linalg.norm(b)
   atol = tol
   assert err < max(rtol, atol)
+
+
+@pytest.mark.parametrize("dtype", np_dtypes)
```
Comment (mganahl) [inline tensornetwork/backends/jax/jax_backend_test.py:774]: seeds everywhere

File: tensornetwork/backends/pytorch/pytorch_backend.py [hunk 2]
```diff
@@ -358,3 +352,78 @@ def matmul(self, tensor1: Tensor, tensor2: Tensor) -> Tensor:
...
+                    sub-arrays from which the diagonals should be taken.
+                    Defaults to first/second axis.
+                    These arguments are not supported by the PyTorch
+                    backend and an error will be raised if they are
+                    specified.
+    Returns:
+      array_of_diagonals: The batched summed diagonals.
+    """
+    if axis1 != -2 or axis2 != -1 or offset != 0:
+      errstr = "offset, axis1, axis2 unsupported by TensorFlow backend."
```
Comment (mganahl) [inline tensornetwork/backends/pytorch/pytorch_backend.py:427]: TensorFlow ->PyTorch

Comment (mganahl) [reply tensornetwork/backends/pytorch/pytorch_backend.py:427]: why can't we support this?

Comment (mganahl) [reply tensornetwork/backends/pytorch/pytorch_backend.py:427]: can you output the actual values of axis1, axis2 and offset?

File: tensornetwork/backends/symmetric/symmetric_backend.py [hunk 3]
```diff
@@ -320,15 +313,35 @@ def broadcast_right_multiplication(self, tensor1: Tensor,
...
+               axis2: int = -1) -> Tensor:
+    if axis1 != -2 or axis2 != -1 or offset != 0:
+      errstr = "offset, axis1, axis2 unsupported by Symmetric backend."
+      raise NotImplementedError(errstr)
+    return self.bs.diag(tensor)
+  
+  def trace(self, tensor: Tensor, offset: int = 0, axis1: int = -2,
+            axis2: int = -1) -> Tensor:
+    # Default np.trace uses first two axes.
+    if axis1 != -2 or axis2 != -1 or offset != 0:
```
Comment (mganahl) [inline tensornetwork/backends/symmetric/symmetric_backend.py:344]: block-sparse trace can be done for arbitrary values of axis

File: tensornetwork/backends/tensorflow/tensorflow_backend.py [hunk 4]
```diff
@@ -267,3 +262,82 @@ def matmul(self, tensor1: Tensor, tensor2: Tensor) -> Tensor:
...
+
+                    These arguments are not supported in the TensorFlow
+                    backend and an error will be raised if they are
+                    specified.
+    Returns:
+      array_of_diagonals: A dim = min(1, tensor.ndim - 2) tensor storing
+                          the batched diagonals.
+    """
+    if axis1 != -2 or axis2 != -1:
+      errstr = "axis1, axis2 unsupported by TensorFlow backend."
```
Comment (mganahl) [inline tensornetwork/backends/tensorflow/tensorflow_backend.py:297]: add actual values of axis1, axis2, and offset

File: tensornetwork/backends/tensorflow/tensorflow_backend.py [hunk 5]
```diff
@@ -267,3 +262,82 @@ def matmul(self, tensor1: Tensor, tensor2: Tensor) -> Tensor:
...
+      axis1, axis2: Axis to be used as the first/second axis of the 2D
+                    sub-arrays from which the diagonals should be taken.
+                    Defaults to first/second axis.
+                    These arguments are not supported in the TensorFlow
+                    backend and an error will be raised if they are
+                    specified.
+    Returns:
+      array_of_diagonals: The batched summed diagonals.
+    """
+    if axis1 != -2 or axis2 != -1 or offset != 0:
```
Comment (mganahl) [inline tensornetwork/backends/tensorflow/tensorflow_backend.py:340]: we could support that

File: tensornetwork/backends/tensorflow/tensorflow_backend_test.py [hunk 6]
```diff
@@ -513,3 +498,48 @@ def test_matmul():
   actual = backend.matmul(a, b)
   expected = np.matmul(t1, t2)
   np.testing.assert_allclose(expected, actual)
+
+
+@pytest.mark.parametrize("dtype", tf_dtypes)
```
Comment (mganahl) [inline tensornetwork/backends/tensorflow/tensorflow_backend_test.py:503]: seeds

File: tensornetwork/backends/pytorch/pytorch_backend.py (hunk 2 above)
Comment (alewis) [reply tensornetwork/backends/pytorch/pytorch_backend.py:427]: I couldn't figure out how (to support this)

Comment (alewis) [reply tensornetwork/backends/pytorch/pytorch_backend.py:427]: Anyway, that can be added later

Comment (mganahl) [reply tensornetwork/backends/pytorch/pytorch_backend.py:427]: a naive way is to translate this into an einsum call

Comment (mganahl) [reply tensornetwork/backends/pytorch/pytorch_backend.py:427]: I think that should do 

```python 
N = len(tensor.shape)

if axis1 < 0:
    axis1 = N+axis1
if axis2 < 0:
    axis2 = N+axis2

inds = list(map(chr,range(97, 97+N)))
indsout = [i for n, i in enumerate(inds) if n not in (axis1, axis2)]
inds[axis1] = 'A'
inds[axis2] = 'A'
return torchlib.einsum(''.join(inds) + '->' +''.join(indsout), tensor)
```

Comment (mganahl) [reply tensornetwork/backends/pytorch/pytorch_backend.py:427]: doesn't help with the offset of course

Comment (mganahl) [reply tensornetwork/backends/pytorch/pytorch_backend.py:427]: or you just transpose the corresponding axis to the end and use the existing call, though that will likely be slower

Comment (alewis) [reply tensornetwork/backends/pytorch/pytorch_backend.py:427]: It doesn't work

Comment (alewis) [reply tensornetwork/backends/pytorch/pytorch_backend.py]: Fixed it

**Comment by alewis** [comment]
I'll fix the conflicts after sign and decomp get merged, since these will introduce more conflicts

**Comment by alewis** [comment]
Should be ready

File: tensornetwork/backends/pytorch/pytorch_backend.py [hunk 7]
```diff
@@ -359,6 +353,98 @@ def matmul(self, tensor1: Tensor, tensor2: Tensor) -> Tensor:
...
+      axis1, axis2: Axis to be used as the first/second axis of the 2D
+                    sub-arrays from which the diagonals should be taken.
+                    Defaults to first/second axis.
+                    These arguments are not supported by the PyTorch
+                    backend and an error will be raised if they are
+                    specified.
+    Returns:
+      array_of_diagonals: The batched summed diagonals.
+    """
+    if offset != 0:
```
Comment (mganahl) [inline tensornetwork/backends/pytorch/pytorch_backend.py:426]: We could actually also have an offset by using contraction with an offset identity matrix.

Comment (mganahl) [reply tensornetwork/backends/pytorch/pytorch_backend.py:426]: Let's leave it for later though
